File: res/shuffle.cpp
#include <iostream>
#include <vector>
#include <cstdlib>
#include <ctime>
#include <set>
#include <numeric>

std::vector<int> partial_shuffle(std::vector<int>& A, int n, int c) {
    std::srand(std::time(0));
    for (int i = 0; i < c; ++i) {
        int j = i + std::rand() % (n - i);
        std::swap(A[i], A[j]);
    }
    return std::vector<int>(A.begin(), A.begin() + c);
}

std::set<int> select_slots(std::vector<int>& y, int n) {
    double p = y[0] / std::accumulate(y.begin(), y.end(), 0.0);
    int c = static_cast<int>(std::ceil(p * n));
    std::vector<int> A(n);
    for (int i = 0; i < n; ++i) {
        A[i] = i + 1;
    }
    partial_shuffle(A, n, c);
    std::set<int> s;
    for (int i = 0; i < c; ++i) {
        s.insert(A[i]);
    }
    return s;
}

int main() {
    std::vector<int> y = {10, 6, 3, 2, 5};  // Example queue lengths including self node
    int n = y.size();
    std::set<int> selected_slots = select_slots(y, n);

    std::cout << "Selected transmission slots: ";
    for (int slot : selected_slots) {
        std::cout << slot << " ";
    }
    std::cout << std::endl;

    return 0;
}

File: res/shuffle.py
import random
import math

def partial_shuffle(A, n, c):
    """
    Perform a partial Fisher-Yates shuffle to select c time slots from n.
    
    Parameters:
    A (list): List of time slots
    n (int): Number of nodes requiring transmission among neighbors
    c (int): Number of selected slots

    Returns:
    list: A list of c selected time slots
    """
    for i in range(c):
        j = random.randint(i, n - 1)
        A[i], A[j] = A[j], A[i]
    return A[:c]

'''
# Example usage:
time_slots = list(range(1, 11))  # Example list of time slots
n = len(time_slots)
c = 5  # Number of slots to select
selected_slots = partial_shuffle(time_slots, n, c)
print("Selected time slots:", selected_slots)
'''

def select_slots(y, n):
    p = y[0] / sum(y)
    c = math.ceil(p * n)
    A = list(range(1, n + 1))
    partial_shuffle(A, n, c)
    s = set()
    for i in range(c):
        s.add(A[i])
    return s

# Example usage
y = [10, 1, 1, 1, 1]  # Example queue lengths including self node
n = len(y)
selected_slots = select_slots(y, n)
print("Selected transmission slots:", selected_slots)